### PR TITLE
feat: initial chain setup

### LINKED ISSUE #15: Setup initial chain parameters
Check if AURA is possible
BSX Token setup
Accounts creation
Treasury

### PR DESCRIPTION
**Pull request by martinfridrich**
closes #15

### CODE REVIEW

File: node/src/chain_spec.rs [hunk 1]
```diff
@@ -206,6 +249,8 @@ fn parachain_genesis(
 	endowed_accounts: Vec<AccountId>,
 	_enable_println: bool,
 	parachain_id: ParaId,
+	council_members: Vec<AccountId>,
+	tech_committe_members: Vec<AccountId>,
```
Comment (green-jay) [inline node/src/chain_spec.rs:253]: typo

File: runtime/src/lib.rs [hunk 2]
```diff
@@ -362,6 +371,170 @@ impl parachain_info::Config for Runtime {}
...
+impl pallet_democracy::Config for Runtime {
+	type Proposal = Call;
+	type Event = Event;
+	type Currency = Balances;
+	type EnactmentPeriod = EnactmentPeriod;
+	type LaunchPeriod = LaunchPeriod;
+	type VotingPeriod = VotingPeriod;
+	type MinimumDeposit = MinimumDeposit;
+	/// A straight majority of the council can decide what their next motion is.
+	type ExternalOrigin = frame_system::EnsureOneOf<
```
[the remark this answers is not in the document]
Comment (martinfridrich) [reply runtime/src/lib.rs:396]: what about this? 
``` rust
type EnsureMajorityCouncilOrRoot = frame_system::EnsureOneOf<
	AccountId,
	pallet_collective::EnsureProportionAtLeast<_1, _2, AccountId, CouncilCollective>,
	frame_system::EnsureRoot<AccountId>,
>;
type EnsureUnanimousCouncilOrRoot = frame_system::EnsureOneOf<
	AccountId,
	pallet_collective::EnsureProportionAtLeast<_1, _1, AccountId, CouncilCollective>,
	frame_system::EnsureRoot<AccountId>,
>;
type EnsureSuperMajorityCouncilOrRoot = frame_system::EnsureOneOf<
	AccountId,
	pallet_collective::EnsureProportionAtLeast<_2, _3, AccountId, CouncilCollective>,
	frame_system::EnsureRoot<AccountId>,
>;
type EnsureSuperMajorityTechCommitteeOrRoot = frame_system::EnsureOneOf<
	AccountId,
	pallet_collective::EnsureProportionAtLeast<_2, _3, AccountId, TechnicalCollective>,
	frame_system::EnsureRoot<AccountId>,
>;
type EnsureUnanimousTechCommitteOrRoot = frame_system::EnsureOneOf<
	AccountId,
	pallet_collective::EnsureProportionAtLeast<_1, _1, AccountId, TechnicalCollective>,
	frame_system::EnsureRoot<AccountId>,
>;
impl pallet_democracy::Config for Runtime {
	type Proposal = Call;
	type Event = Event;
	type Currency = Balances;
	type EnactmentPeriod = EnactmentPeriod;
	type LaunchPeriod = LaunchPeriod;
	type VotingPeriod = VotingPeriod;
	type MinimumDeposit = MinimumDeposit;
	/// A straight majority of the council can decide what their next motion is.
	type ExternalOrigin = EnsureMajorityCouncilOrRoot;
	/// A majority can have the next scheduled referendum be a straight majority-carries vote
	type ExternalMajorityOrigin = EnsureMajorityCouncilOrRoot;
	/// A unanimous council can have the next scheduled referendum be a straight default-carries
	/// (NTB) vote.
	type ExternalDefaultOrigin = EnsureUnanimousCouncilOrRoot;
	/// Two thirds of the technical committee can have an ExternalMajority/ExternalDefault vote
	/// be tabled immediately and with a shorter voting/enactment period.
	type FastTrackOrigin = EnsureSuperMajorityTechCommitteeOrRoot;
	type InstantOrigin = EnsureUnanimousTechCommitteOrRoot;
	type InstantAllowed = InstantAllowed;
	type FastTrackVotingPeriod = FastTrackVotingPeriod;
	// To cancel a proposal which has been passed, 2/3 of the council must agree to it.
	type CancellationOrigin = EnsureSuperMajorityCouncilOrRoot;
	// To cancel a proposal before it has been passed, the technical committee must be unanimous or
	// Root must agree.
	type CancelProposalOrigin = EnsureUnanimousTechCommitteOrRoot;
	type BlacklistOrigin = EnsureRoot<AccountId>;
	// Any single technical committee member may veto a coming council proposal, however they can
	// only do it once and it lasts only for the cooloff period.
	type VetoOrigin = pallet_collective::EnsureMember<AccountId, TechnicalCollective>;
	type CooloffPeriod = CooloffPeriod;
	type PreimageByteDeposit = PreimageByteDeposit;
	type OperationalPreimageOrigin = pallet_collective::EnsureMember<AccountId, CouncilCollective>;
	type Slash = Treasury;
	type Scheduler = Scheduler;
	type PalletsOrigin = OriginCaller;
	type MaxVotes = MaxVotes;
	type WeightInfo = ();
	type MaxProposals = MaxProposals;
}
```

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -123,6 +127,11 @@ pub const EPOCH_DURATION_IN_BLOCKS: u32 = 10 * MINUTES;
 pub const MINUTES: BlockNumber = 60_000 / (MILLISECS_PER_BLOCK as BlockNumber);
 pub const HOURS: BlockNumber = MINUTES * 60;
 pub const DAYS: BlockNumber = HOURS * 24;
+pub const FORTUNE: Balance = u128::MAX;
+pub const BSX: Balance = 1_000_000_000_000;
+pub const DOLLARS: Balance = BSX * 10; // 10 BSX ~= 1 $
```
Comment (lumir-mrkva) [inline runtime/src/lib.rs:132]: HDX = 10 BSX so it should be 100 BSX for $1

File: runtime/src/lib.rs [hunk 4]
```diff
@@ -362,6 +371,168 @@ impl parachain_info::Config for Runtime {}
...
+	type MotionDuration = TechnicalMotionDuration;
+	type MaxProposals = TechnicalMaxProposals;
+	type MaxMembers = TechnicalMaxMembers;
+	type DefaultVote = pallet_collective::PrimeDefaultVote;
+	type WeightInfo = ();
+}
+
+parameter_types! {
+	pub const ProposalBond: Permill = Permill::from_percent(5);
+	pub const ProposalBondMinimum: Balance = FORTUNE;
```
Comment (lumir-mrkva) [inline runtime/src/lib.rs:493]: this means you cannot actually propose?

**Review by lumir-mrkva — COMMENTED** [review]
i think that BSX governance should be more fast and open. it would be advisable to test params on snakenet 1st though

File: runtime/src/lib.rs (hunk 4 above)
Comment (lumir-mrkva) [reply runtime/src/lib.rs:493]: set reasonably high value equaling $100

File: node/src/chain_spec.rs [hunk 5]
```diff
@@ -68,23 +79,43 @@ pub fn roccocco_parachain_config(para_id: ParaId) -> Result<ChainSpec, String> {
 				wasm_binary,
 				// Sudo account
 				hex!["30035c21ba9eda780130f2029a80c3e962f56588bc04c36be95a225cb536fb55"].into(),
```
Comment (jak-pan) [inline node/src/chain_spec.rs:81]: What are these addresses? Don't we want alice/bob for testnet?

File: node/src/command.rs [hunk 6]
```diff
@@ -38,7 +38,7 @@ use std::{io::Write, net::SocketAddr};
 fn load_spec(id: &str, para_id: ParaId) -> std::result::Result<Box<dyn sc_service::ChainSpec>, String> {
 	Ok(match id {
 		"dev" => Box::new(chain_spec::parachain_development_config(para_id)?),
-		"" => Box::new(chain_spec::roccocco_parachain_config(para_id)?),
+		"testnet" => Box::new(chain_spec::testnet_parachain_config(para_id)?),
```
Comment (lumir-mrkva) [inline node/src/command.rs:41]: is this change necessary? you have to change local startup procedure as well and polkadot launch config files

File: node/src/chain_spec.rs (hunk 5 above)
Comment (jak-pan) [reply node/src/chain_spec.rs:81]: Ah might have misunderstood the use of this config. What's the purpose? Real testing?

File: runtime/src/lib.rs [hunk 7]
```diff
@@ -362,6 +371,170 @@ impl parachain_info::Config for Runtime {}
...
+	type Slash = Treasury;
+	type Scheduler = Scheduler;
+	type PalletsOrigin = OriginCaller;
+	type MaxVotes = MaxVotes;
+	type WeightInfo = ();
+	type MaxProposals = MaxProposals;
+}
+
+parameter_types! {
+	pub const CouncilMotionDuration: BlockNumber = 5 * DAYS;
```
[the remark this answers is not in the document]
Comment (jak-pan) [reply runtime/src/lib.rs:455]: Can't go much faster here since you still need to sync across multiple people and timezones...

File: node/src/chain_spec.rs (hunk 5 above)
Comment (martinfridrich) [reply node/src/chain_spec.rs:81]: yes, this is config for public/real testing(westend/rococo).

File: node/src/command.rs (hunk 6 above)
Comment (martinfridrich) [reply node/src/command.rs:41]: I think it's better if you have to specify a chain. You will know what you are running without looking into code to find out what chain is under `""`.  Polkadot launch running `--chain=local`. It should not be affected by this change. I can revert it back to `""` if we want

File: node/src/chain_spec.rs [hunk 8]
```diff
@@ -1,18 +1,29 @@
...
+use sp_core::{crypto::UncheckedInto, sr25519, Pair, Public};
 use sp_runtime::traits::{IdentifyAccount, Verify};
 
+const TOKEN_DECIMALS: u8 = 12;
+const TOKEN_SYMBOL: &str = "BSX";
+const PROTOCOL_ID: &str = "bsx";
+// The URL for the telemetry server.
+const TELEMETRY_URLS: [&str; 2] = [
+	"wss://telemetry.polkadot.io/submit/",
+	"wss://telemetry.hydradx.io:9000/submit/",
```
[the remark this answers is not in the document]
Comment (martinfridrich) [reply node/src/chain_spec.rs:24]: sorry, I don't understand what you mean

File: node/src/command.rs (hunk 6 above)
Comment (lumir-mrkva) [reply node/src/command.rs:41]: ok, it is not issue for polkadot-launch afterall

File: runtime/src/lib.rs [hunk 9]
```diff
@@ -384,6 +400,168 @@ impl parachain_info::Config for Runtime {}
...
+	type WeightInfo = ();
+}
+
+parameter_types! {
+	pub const ProposalBond: Permill = Permill::from_percent(5);
+	pub const ProposalBondMinimum: Balance = 100 * DOLLARS;
+	pub const SpendPeriod: BlockNumber = DAYS;
+	pub const Burn: Permill = Permill::from_percent(0);
+	pub const TreasuryPalletId: PalletId = PalletId(*b"py/trsry");
+	pub const MaxApprovals: u32 =  100; //TODO: someone should check this value
```
Comment (lumir-mrkva) [inline runtime/src/lib.rs:526]: final value?

File: runtime/src/lib.rs [hunk 10]
```diff
@@ -384,6 +400,168 @@ impl parachain_info::Config for Runtime {}
...
+	type Event = Event;
+	type MotionDuration = CouncilMotionDuration;
+	type MaxProposals = CouncilMaxProposals;
+	type MaxMembers = CouncilMaxMembers;
+	type DefaultVote = pallet_collective::PrimeDefaultVote;
+	type WeightInfo = ();
+}
+
+parameter_types! {
+	pub const TechnicalMotionDuration: BlockNumber = 7 * DAYS;
```
Comment (lumir-mrkva) [inline runtime/src/lib.rs:503]: any reason for a different length than the council motion?

**Review by lumir-mrkva — APPROVED** [review]

**Comment by lumir-mrkva** [comment]
LGTM

File: node/src/chain_spec.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jak-pan) [reply node/src/chain_spec.rs:24]: telemetry. If we should have separate for basilisk but I guess not needed for now

File: runtime/src/lib.rs (hunk 9 above)
Comment (jak-pan) [reply runtime/src/lib.rs:526]: It's ok

File: runtime/src/lib.rs (hunk 10 above)
Comment (jak-pan) [reply runtime/src/lib.rs:503]: I'd probably make this shorter too

Comment (jak-pan) [reply runtime/src/lib.rs:503]:
```suggestion
	pub const TechnicalMotionDuration: BlockNumber = 5 * DAYS;
```

**Review by jak-pan — CHANGES_REQUESTED** [review]
Epoch time should not be 10 minutes. Other than that LGTM

**Comment by jak-pan** [comment]
> Epoch time should not be 10 minutes. Other than that LGTM

Is it actually used anywhere ? @martinfridrich

**Review by jak-pan — APPROVED** [review]

**Comment by jak-pan** [comment]
@lumir-mrkva @fakirAyoub guys do you see why we don't have workflow run on the force-push?

**Comment by lumir-mrkva** [comment]
@jak-pan not sure i understand i see the commit "remove epoch" and it has workflow on it. anyway i wouldn't recommend to use force push in any case

File: node/src/chain_spec.rs [hunk 11]
```diff
@@ -251,6 +296,15 @@ fn parachain_genesis(
...
+		pallet_collective_Instance1: CouncilConfig {
+			members: council_members,
+			phantom: Default::default(),
+		},
+		pallet_collective_Instance2: TechnicalCommitteeConfig {
+			members: tech_committee_members,
+			phantom: Default::default(),
+		},
 		parachain_info: ParachainInfoConfig { parachain_id },
 		pallet_faucet: FaucetConfig {
```
Comment (lumir-mrkva) [inline node/src/chain_spec.rs:309]: faucet should be removed

File: runtime/src/lib.rs [hunk 12]
```diff
@@ -153,6 +160,13 @@ impl Filter<Call> for BaseFilter {
...
+            //TODO: validate this
+            | Call::Democracy(_)
+            | Call::Scheduler(_)
+            | Call::Council(_)
+            | Call::TechnicalCommittee(_)
+            | Call::Treasury(_)
+            //TODO: validate this end
 			| Call::Sudo(_) => true,
 
 			Call::XYK(_)
```
Comment (lumir-mrkva) [inline runtime/src/lib.rs:172]: just filter out XYK, everything else should be allowed

File: runtime/src/lib.rs [hunk 13]
```diff
@@ -153,6 +160,13 @@ impl Filter<Call> for BaseFilter {
 			| Call::Timestamp(_)
 			| Call::RandomnessCollectiveFlip(_)
 			| Call::ParachainSystem(_)
+            //TODO: validate this
```
Comment (lumir-mrkva) [inline runtime/src/lib.rs:163]: sure allow everything, we dont have to freeze balances on basilisk

**Review by lumir-mrkva — CHANGES_REQUESTED** [review]
one more thing ...

**Review by lumir-mrkva — APPROVED** [review]
